### PR TITLE
URL encode datasources, task ids, authenticator names.

### PR DESCRIPTION
**Pull request by gianm**
Fixes various bugs related to objects with characters in their names
that must be URL encoded (such as spaces). The one that made me notice
this was that the task log and shutdown APIs don't work when task IDs
have spaces in them.

Includes a fix for a bug in which the router does not forward encoded
paths properly (it would double-encode them).

### CODE REVIEW

**Comment by jihoonson** [comment]
I think this is a very common mistake, so that we need another way to prevent such mistakes. I suggest to add a sort of `UrlBuilder` which enforces encoding to input elements if needed.

**Comment by himanshug** [comment]
nasty bug!

> I think this is a very common mistake, so we need another way to prevent such mistakes. I suggest to add a sort of UrlBuilder which enforces encoding to input elements if needed.

How would you check the "if needed" ?

**Comment by himanshug** [comment]
looks like most of the fixes are in the calls to makeWorkerUrl(..), is it possible to put the fix inside impl of that method instead ?

**Comment by gianm** [comment]
@himanshug, re:

> looks like most of the fixes are in the calls to makeWorkerUrl(..), is it possible to put the fix inside impl of that method instead ?

I made that change and re-pushed. I also consolidated the two makeWorkerURL methods into one helper method in TaskRunnerUtils.

@jihoonson, re:

> I think this is a very common mistake, so we need another way to prevent such mistakes. I suggest to add a sort of UrlBuilder which enforces encoding to input elements if needed.

I considered doing this, but then we have a new problem of how to get callers to use UrlBuilder rather than making their own URLs. So I ended up just adding a urlEncode method to StringUtils and calling that where appropriate. (Note: it isn't always appropriate, since sometimes the format parameter is a literal path string like `"foo/bar"` and in that case we should _not_ url encode.)

**Comment by gianm** [comment]
I've pushed a fix for the tests.

**Comment by gianm** [comment]
Fixed conflicts.

**Comment by jihoonson** [comment]
Please check the build failure.

```
[ERROR] Forbidden field access: java.lang.System#out [prints to System.out; should only be used for debugging, not in production code]
[ERROR]   in io.druid.query.topn.types.NumericTopNColumnSelectorStrategy (NumericTopNColumnSelectorStrategy.java:153)
[ERROR] Forbidden method invocation: java.lang.String#format(java.lang.String,java.lang.Object[]) [Uses default locale]
[ERROR]   in io.druid.query.topn.types.NumericTopNColumnSelectorStrategy (NumericTopNColumnSelectorStrategy.java:153)
[ERROR] Forbidden field access: java.lang.System#out [prints to System.out; should only be used for debugging, not in production code]
[ERROR]   in io.druid.query.topn.TopNLexicographicResultBuilder (TopNLexicographicResultBuilder.java:79)
[ERROR] Scanned 1596 (and 630 related) class file(s) for forbidden API invocations (in 1.63s), 3 error(s).
```

**Comment by fjy** [comment]
@gianm this is having a hard time passing integration tests

**Comment by gianm** [comment]
I wasn't sure if it was timing out (we are pretty close to the timeout) or if there was something wrong with the patch that affects the integration tests. I'll take a closer look soon.

**Comment by gianm** [comment]
I added some fixes for the router that might have been affecting the integration tests. It was double-encoding any encoded paths that it received from a client, meaning it would garble those URLs.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunnerUtils.java [hunk 1]
```diff
@@ -89,4 +98,20 @@ public void run()
       }
     }
   }
+
+  public static URL makeWorkerURL(Worker worker, String pathFormat, Object... pathParams)
```
Comment (jihoonson) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunnerUtils.java]: Would you please add a unit test?

File: server/src/main/java/org/apache/druid/server/JettyUtils.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+public class JettyUtils
+{
+  /**
+   * Concatenate URI parts, in a way that is useful for proxy servlets.
+   *
+   * @param base               base part of the uri, like http://example.com (no trailing slash)
+   * @param encodedPath        encoded path, like you would get from HttpServletRequest's getRequestURI
+   * @param encodedQueryString encoded query string, like you would get from HttpServletRequest's getQueryString
+   */
+  public static String concatenateForRewrite(
```
Comment (jihoonson) [inline server/src/main/java/org/apache/druid/server/JettyUtils.java]: Would you please add a unit test?

**Comment by jihoonson** [comment]
Also please check the travis fail. Looks legit.

```
[ERROR] Tests run: 8, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1,646.446 s <<< FAILURE! - in TestSuite
[ERROR] testIndexData(org.apache.druid.tests.indexer.ITParallelIndexTest)  Time elapsed: 242.129 s  <<< FAILURE!
org.apache.druid.java.util.common.ISE: Error while making request to url[http://127.0.0.1:8081/druid/coordinator/v1/datasources/wikipedia_parallel_index_test/intervals] status[204 No Content] content[]
	at org.apache.druid.tests.indexer.ITParallelIndexTest.testIndexData(ITParallelIndexTest.java:44)
```

**Comment by fjy** [comment]
@gianm I think this is failing integration tests

**Comment by gianm** [comment]
The integration test issue seems to be due to some double-encoding happening (the URI constructor encodes its parameters, URL doesn't). Will sort that out.

**Comment by gianm** [comment]
I fixed the double-encoding and I think this is good to go now.

**Pull request merged**